tensorflow/python/ops/linalg_grad.py

```diff
@@ -54,6 +54,65 @@ def _MatrixDeterminantGrad(op, grad):
                                                    0))
   return multipliers * a_adj_inv
 
+@ops.RegisterGradient("MatrixSquareRoot")
+def _MatrixSquareRootGrad(op, grad):
+  """Gradient for MatrixSquareRoot."""
+
+  # Let A be an m x m square matrix (or batch of matrices)
+  # Let R = sqrtm(A)
+  # By definition, A = RR
+  # Take the differential: dA = d(RR) = RdR + dRR
+  # Solve the resulting Sylvester equation for dR
+
+  # Used to find Kronecker products within the Sylvester equation
+  def _KroneckerProduct(mat1, mat2):
+    """Computes the Kronecker product of two matrices."""
+    m1, n1 = mat1.get_shape().as_list()
+    mat1_rsh = array_ops.reshape(mat1, [m1, 1, n1, 1])
+    m2, n2 = mat2.get_shape().as_list()
+    mat2_rsh = array_ops.reshape(mat2, [1, m2, 1, n2])
+    return array_ops.reshape(mat1_rsh * mat2_rsh, [m1 * m2, n1 * n2])
+
+  sqrt_batch = op.outputs[0] # R
+  shape_batch = sqrt_batch.get_shape()
+  order = shape_batch.as_list()[-1] # m
+  identity = _linalg.eye(order, dtype=sqrt_batch.dtype) # m x m identity matrix
+  shape_matrix = [order, order] # Tensor may be a batch of matrices of shape m x m
+
+  # Flatten batches containing R and dA
+  flat_sqrt = array_ops.reshape(sqrt_batch, [-1])
+  flat_grad = array_ops.reshape(grad, [-1])
+
+  # Split flattened batches into m x m matrices
+  num_elements = flat_sqrt.get_shape().as_list()[-1]
+  num_splits = int(num_elements / (order * order))
+  split_sqrt = array_ops.split(flat_sqrt, num_splits)
+  split_grad = array_ops.split(flat_grad, num_splits)
+
+  matrix_gradients = [] # Raw gradients of all m x m matrices
+  for flat_sqrt_matrix, flat_grad_matrix in zip(split_sqrt, split_grad):
+    sqrt_matrix = array_ops.reshape(flat_sqrt_matrix, shape_matrix) # m x m matrix R
+    grad_matrix = array_ops.reshape(flat_grad_matrix, shape_matrix) # m x m matrix dA
+
+    # The transpose of R is taken in the k1 term instead of k2 in
+    # order to prevent redundant transposition of R (i.e. (R')' = R)
+    sqrt_matrix_transpose = array_ops.matrix_transpose(sqrt_matrix)
+    k1 = _KroneckerProduct(identity, sqrt_matrix_transpose)
+    k2 = _KroneckerProduct(sqrt_matrix, identity)
+
+    # Solve for vec(dR) by vectorizing dA
+    inv_ksum = _linalg.inv(math_ops.add(k1, k2))
```

Comment:
Instead of forming the inverse and multiplying by it, can you write this in terms of linalg.matrix_solve?